basis/http/server/static/static.factor

```diff
@@ -58,18 +58,98 @@ TUPLE: file-responder root hook special index-names allow-listings ;
 
 \ serve-file NOTICE add-input-logging
 
-: file>html ( name -- xml )
-    dup link-info directory? [ "/" append ] when
-    dup [XML <li><a href=<->><-></a></li> XML] ;
+:: file-html-template ( href size modified -- xml )
+    [XML
+        <tr>
+            <td><a href=<-href->><-href-></a></td>
+            <td align="right"><-modified-></td>
+            <td align="right"><-size-></td>
+        </tr>
+    XML] ;
 
-: directory>html ( path -- xml )
-    [ file-name ]
-    [ drop f ]
+: file>html ( name infos -- xml )
     [
-        [ file-name ] [ [ natural-sort [ file>html ] map ] with-directory-files ] bi
-        [XML <h1><-></h1> <ul><-></ul> XML]
-    ] tri
-    simple-page ;
+        dup directory?
+        [ drop "/" append "-" ]
+        [ size>> number>string ] if
+    ] [ modified>> ] bi file-html-template ;
+
+: parent-dir-link ( -- xml )
+    "../" "" "" file-html-template ;
+
+: ?parent-dir-link ( -- xml/f )
+    url get [ path>> "/" = [ "" ] [ parent-dir-link ] if ] [ "" ] if* ;
+
+: listing-title ( -- title )
+    url get [ path>> "Index of " prepend ] [ "" ] if* ;
+
+:: listing-html-template ( title listing ?parent CO-N CO-M CO-S -- xml )
+    [XML <h1><-title-></h1>
+        <table>
+            <tr>
+                <th><a href=<-CO-N->>Name</a></th>
+                <th><a href=<-CO-M->>Last modified</a></th>
+                <th><a href=<-CO-S->>Size</a></th>
+            </tr>
+            <tr><th colspan="5"><hr/></th></tr>
+            <-?parent->
+            <-listing->
+            <tr><th colspan="5"><hr/></th></tr>
+        </table>
+    XML] ;
+
+: sort-column ( -- column ) params get "C" of "N" or ;
+
+: sort-order ( -- order ) params get "O" of "A" or ;
+
+: sort-asc? ( -- ? ) sort-order "A" = ;
+
+: toggle-order ( order -- order' ) "A" = "D" "A" ? ;
+
+: ?toggle-sort-order ( col current-col -- order )
+    = [ sort-order toggle-order ] [ "A" ] if ;
+
+: sort-orders ( -- CO-N CO-M CO-S )
+    "N" "M" "S" sort-column [
+        [ drop "?C=" ";O=" surround ]
+        [ ?toggle-sort-order ] 2bi append
+    ] curry tri@ ;
+
+: listing-sort-with ( seq quot: ( elt -- key ) -- sortedseq )
+    sort-with sort-asc? [ reverse ] unless ; inline
+
+: sort-with-name ( {file,info} -- sorted )
+    [ first ] listing-sort-with ;
+
+: sort-with-modified ( {file,info} -- sorted )
+    [ second modified>> ] listing-sort-with ;
+
+: size-without-directories ( info -- size )
+    dup directory? [ drop -1 ] [ size>> ] if ;
+
+: sort-with-size ( {file,info} -- sorted )
+    [ second size-without-directories ] listing-sort-with ;
```

Comment:
Another idea might be to make a tuple that has the ``{ name modified size }`` slots and then use ``sorting.slots`` to do a secondary sort when sorting by size or modified in case of identical size/modified slots.

```{ "size" "name" } sort-by```

```{ "size" "modified" } sort-by```